File: examples/csharp/MyPaymentService.Core/Application/UseCases/GetPaymentUseCase.cs
using MyPaymentService.Core.Application.Contracts;
using MyPaymentService.Core.Capabilities;

namespace MyPaymentService.Core.Application.UseCases;

public class GetPaymentUseCase(IPaymentRepository paymentRepository)
{
    public GetPaymentResponse? Execute(GetPaymentRequest request)
    {
        var payment = paymentRepository.GetById(request.PaymentId);

        if (payment == null)
            return null;

        return new GetPaymentResponse(
            payment.Id,
            payment.Amount,
            payment.MerchantId,
            payment.Status,
            payment.CreatedAt,
            payment.DeclineReason);
    }
}

File: examples/cpp/mypaymentservice_core/application/use_cases/get_payment_use_case.cpp
#include "mypaymentservice_core/application/use_cases/get_payment_use_case.h"

namespace mypaymentservice::core::application::use_cases {

GetPaymentUseCase::GetPaymentUseCase(capabilities::IPaymentRepository* repository)
    : repository_(repository) {
}

std::optional<contracts::GetPaymentResponse>
GetPaymentUseCase::Execute(const std::string& payment_id) {
    return repository_->GetById(payment_id);
}

} // namespace mypaymentservice::core::application::use_cases
